kpatch-build/create-klp-module.c

```diff
@@ -0,0 +1,590 @@
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA,
+ * 02110-1301, USA.
+ */
+
+#include <string.h>
+#include <stdlib.h>
+#include <libgen.h>
+#include <argp.h>
+
+#include "log.h"
+#include "kpatch-elf.h"
+
+/* For log.h */
+char *childobj;
+enum loglevel loglevel = NORMAL;
+
+/* klp_* structs and lists used for temporary sorting/bookkeeping */
+struct klp_relasec {
+	struct section *sec;
+	struct section *base;
+	struct list_head list;
+	char *objname;
+};
+
+struct klp_symbol {
+	struct symbol *sym;
+	char *oldname;
+	char *objname;
+	unsigned long pos;
+	struct list_head list;
+};
+
+struct klp_object {
+	char *name;
+	struct list_head arch_entries;
+	struct list_head list;
+};
+
+struct klp_arch_entry {
+	struct section *sec;
+	struct list_head list;
+};
+
+static LIST_HEAD(klp_relasecs);
+static LIST_HEAD(klp_symbols);
+static LIST_HEAD(klp_objects);
+
+static struct symbol *kpatch_add_ksym_to_symbols(struct kpatch_elf *kelf,
+						 struct section *ksymsec,
+						 char *strings, int offset)
+{
+	struct kpatch_symbol *ksyms, *ksym;
+	struct klp_symbol *ks;
+	struct symbol *sym;
+	struct rela *rela;
+	char *objname, *name, *newname;
+	char pos[32];
+	int len, index;
+
+	ksyms = ksymsec->data->d_buf;
+	index = offset / sizeof(*ksyms);
+	ksym = &ksyms[index];
+
+	/* Get name of ksym */
+	rela = find_rela_by_offset(ksymsec->rela, offset + offsetof(struct kpatch_symbol, name));
+	if (!rela)
+		ERROR("name of ksym not found?");
+
+	name = strdup(strings + rela->addend);
+	if (!name)
+		ERROR("strdup");
+
+	/* Get objname of ksym */
+	rela = find_rela_by_offset(ksymsec->rela, offset + offsetof(struct kpatch_symbol, objname));
+	if (!rela)
+		ERROR("objname of ksym not found?");
+
+	objname = strdup(strings + rela->addend);
+	if (!objname)
+		ERROR("strdup");
+
+	/* Look for an already alloc'd symbol */
+	list_for_each_entry(ks, &klp_symbols, list) {
+		if (!strcmp(ks->oldname, name) &&
+		    !strcmp(ks->objname, objname) &&
+		    ks->pos == ksym->pos)
+			return ks->sym;
+	}
+
+	len = snprintf(pos, 32, "%lu", ksym->pos);
+	if (len < 0)
+		ERROR("snprintf");
+
+	/* .klp.sym.objname.name,pos */
+	len = strlen(KLP_SYM_PREFIX) + strlen(objname) + \
+	      strlen(name) + strlen(pos) + 3;
+	newname = malloc(len);
+	if (!newname)
+		ERROR("malloc");
+
+	len = snprintf(newname, len, KLP_SYM_PREFIX "%s.%s,%s", objname, name, pos);
+	if (len < 0)
+		ERROR("snprintf");
+
+	ALLOC_LINK(sym, &kelf->symbols);
+	sym->name = newname;
+	sym->type = ksym->type;
+	sym->bind = ksym->bind;
+	/*
+	 * Note that st_name will be set in kpatch_create_strtab(),
+	 * and sym->index is set in kpatch_reindex_elements()
+	 */
+	sym->sym.st_shndx = SHN_LIVEPATCH;
+	sym->sym.st_info = GELF_ST_INFO(sym->bind, sym->type);
+
+	ALLOC_LINK(ks, &klp_symbols);
+	ks->sym = sym;
+	ks->pos = ksym->pos;
+	ks->oldname = name;
+	ks->objname = objname;
+
+	return sym;
+}
+
+static struct section *kpatch_find_klp_relasec(struct kpatch_elf *kelf,
```

Comment:
Yeah, the 'elf_' prefix is the same namespace used by libelf.h, and might confuse things.  Then again, I don't like 'kelf_' either because ideally our elf code should have no knowledge of kpatch. It should really be a layer above libelf to make it easier to interact with it.

FWIW, I have a very similar library in upstream Linux in tools/objtool/elf.{ch}, though it has diverged a bit -- it has some hashes for fast lookups.  Since I'm using it in two projects, and others might find it useful, it might be nice to move that code to another project someday.  No idea what to call it, or prefix the code with, though.

But anyway, such changes should indeed probably go into a separate PR.